src/graphql/utils.ts

```diff
@@ -1,24 +1,132 @@
+import type {
+  DocumentNode,
+  FragmentDefinitionNode,
+  OperationDefinitionNode,
+  SelectionNode,
+} from 'graphql';
+
 import { debug } from '../helpers';
-import { Entity } from '../types';
+import type { Entity, GraphQLParams } from '../types';
+
+const COLLECTION_SUFFIX = 'Collection';
+
+export function buildCollectionName(name: string): string {
+  return `${name}${COLLECTION_SUFFIX}`;
+}
 
-const DEFAULT_CONTENT_TYPE_FIELDS = ['sys', '__typename', 'contentfulMetadata'];
+function getTypeNameForNode(
+  node: FragmentDefinitionNode | OperationDefinitionNode
+): string | undefined {
+  return 'typeCondition' in node ? node.typeCondition.name.value : node.name?.value;
+}
+
+interface GeneratedGraphQLStructure {
+  name: string;
+  alias?: string;
+  __typename: string;
+}
 
-export function logUnrecognizedFields(contentTypeFields: string[], data: Entity): void {
-  const recognized = new Set([...DEFAULT_CONTENT_TYPE_FIELDS, ...contentTypeFields]);
+function detectFieldAlias(
+  selections: Readonly<SelectionNode[]>,
+  typename: string
+): GeneratedGraphQLStructure[] {
+  const generated: GeneratedGraphQLStructure[] = [];
 
-  for (const field of Object.keys(data)) {
-    if (!(recognized.has(field) || recognized.has(removeCollection(field)))) {
-      debug.warn(`Unrecognized field '${field}'. Note that GraphQL aliases are not supported`);
+  for (const selection of selections) {
+    if (selection.kind === 'Field') {
+      generated.push({
+        name: selection.name.value,
+        alias: selection.alias?.value,
+        __typename: typename,
+      });
+
+      if (selection.selectionSet?.selections) {
+        generated.push(
+          ...detectFieldAlias(selection.selectionSet.selections, selection.name.value)
+        );
+      }
     }
   }
+
+  return generated;
 }
 
-const COLLECTION_SUFFIX = 'Collection';
+export function parseGraphQLParams(query: DocumentNode): GraphQLParams {
```

Comment:
I'm a bit confused, could you please explain what this function does?

Reply:
added documentation 👍